server/src/main/java/org/candlepin/async/JobConfig.java

```diff
@@ -26,11 +28,16 @@
 /**
  * The JobConfig object collects the configuration for a job, which can be provided to the
  * JobManager to queue a new instance of that job.
+ *
+ * @param <T>
+ *  the concrete JobConfig class being implemented
  */
-public class JobConfig {
+public class JobConfig<T extends JobConfig> {
```

Comment:
Yeah, it's odd, and has its own downsides (further subclassing can get janky if the subclass implementors don't also propagate the type). The reason this exists in this case is so that we can use the method chaining without needing to worry about the order of the methods. Since setContextOwner comes from JobConfig, without the generic bit, it'll return a JobConfig and require casting to keep going. With the generic, the subclass doesn't need to reimplement it just to adjust the return type.

As far as the practice or what it's called... I've no idea. I just thought to try it. This is done in a few places to various effect (the DTOs is another such place). I can add comments, but I'm not sure what else to add besides the existing Javadoc for it. What would you like to see there?

Reply:
Just that explanation: "The reason this exists is so that we can use the method chaining without needing to worry about the order of the methods. Since setContextOwner comes from JobConfig, without the generic bit, it'll return a JobConfig and require casting to keep going."